### PR TITLE
[FREELDR] Add/Fix FrLdrHeapAlloc() failure handling and related

### PR DESCRIPTION
**Pull request by SergeGautherie**
Rebased patch that used to be attached to Jira, which @ThFabba had a look at 4 years ago.

JIRA issue: [CORE-12802](https://jira.reactos.org/browse/CORE-12802)

NB:
I am not working on added 'FIXME: Close these keys?', for the time being.

### CODE REVIEW

**Review by HBelusca — DISMISSED** [review]
Can you remove all these additional ERRs you've added? FreeLdr is already quite large and has size limitations; no need to make it fatter by adding many other ERRs and whatnot for failures that never happen in real life.

**Comment by SergeGautherie** [comment]
> remove all these additional ERRs

I removed all new `ERR()`, but 1.

> fatter by adding many other ERRs

May be add a same `ERR("FrLdrHeapAlloc() failed\n");` in all these other cases, instead?

**Comment by learn-more** [comment]
> > remove all these additional ERRs
> 
> I removed all new `ERR()`, but 1.
> 
> > fatter by adding many other ERRs
> 
> May be add a same `ERR("FrLdrHeapAlloc() failed\n");` in all these other cases, instead?

And how is that not increasing the size of the binary?

**Comment by SergeGautherie** [comment]
> > May be add a same `ERR("FrLdrHeapAlloc() failed\n");` in all these other cases, instead?
> 
> And how is that not increasing the size of the binary?

Size would increase a bit, but that should add 1 common string only, not many different ones.

File: boot/freeldr/freeldr/lib/peloader.c
Comment (tkreuzer) [inline boot/freeldr/freeldr/lib/peloader.c:632]: Instead of the custom copy routine and saving the buffer, just use RtlCopyMemory.

File: boot/freeldr/freeldr/lib/peloader.c
Comment (HBelusca) [inline boot/freeldr/freeldr/lib/peloader.c:635]:
```suggestion
    RtlCopyMemory(Buffer, Length);
```

File: boot/freeldr/freeldr/lib/peloader.c
Comment (HBelusca) [reply boot/freeldr/freeldr/lib/peloader.c:632]: Actually that won't work, because it's a poor-man ANSI to UNICODE conversion that is being done here. BaseDllName is a (pointer to) ANSI string, while the allocated and initialized buffer is UNICODE (UTF-16).

Comment (SergeGautherie) [reply boot/freeldr/freeldr/lib/peloader.c:632]: Options are: either leave it as is or switch to `UNICODE_STRING` and `RtlAnsiStringToUnicodeString()`, if the latter is wanted.

File: boot/freeldr/freeldr/lib/peloader.c
Comment (SergeGautherie) [inline boot/freeldr/freeldr/lib/peloader.c:631]: Shall I remove zeroing, which seems useless to me?

Comment (HBelusca) [reply boot/freeldr/freeldr/lib/peloader.c:631]: Just keep all that as it is.

File: boot/freeldr/freeldr/lib/peloader.c
Comment (HBelusca) [reply boot/freeldr/freeldr/lib/peloader.c:632]: Freeldr doesn't include the whole RTL and the unicode -> ANSI tables for obvious reasons of size limitations. So just keep the code as it currently is.

File: boot/freeldr/freeldr/arch/i386/pc/pchw.c
Comment (tkreuzer) [inline boot/freeldr/freeldr/arch/i386/pc/pchw.c:294]: Please use RtlZeroMemory consistently

File: boot/freeldr/freeldr/ntldr/wlregistry.c
Comment (tkreuzer) [inline boot/freeldr/freeldr/ntldr/wlregistry.c:542]: This is most likely a fatal failure. Should probably just call FrLdrBugCheck or something, but at least add an ERR() here.

File: boot/freeldr/freeldr/ntldr/wlregistry.c
Comment (tkreuzer) [inline boot/freeldr/freeldr/ntldr/wlregistry.c:548]: Add ERR()

**Review by tkreuzer — APPROVED** [review]
Generally ok, with some nits, I'd appreciate to be fixed.

File: boot/freeldr/freeldr/ntldr/wlregistry.c
Comment (HBelusca) [reply boot/freeldr/freeldr/ntldr/wlregistry.c:548]: Why only there?

File: boot/freeldr/freeldr/arch/i386/pc/pchw.c
Comment (SergeGautherie) [reply boot/freeldr/freeldr/arch/i386/pc/pchw.c:294]: Agreed. I add commits 2 and 3, which I'll merge after approval.

File: boot/freeldr/freeldr/ntldr/wlregistry.c
Comment (SergeGautherie) [reply boot/freeldr/freeldr/ntldr/wlregistry.c:548]: (See my other comment.)

File: boot/freeldr/freeldr/ntldr/wlregistry.c
Comment (SergeGautherie) [reply boot/freeldr/freeldr/ntldr/wlregistry.c:542]: Adding `ERR()`s was previously refused (twice)! Simply returning matches the other failure cases.
If you want more, in this PR or a follow-up, please get explicit approval from @HBelusca and @learn-more.

**Comment by SergeGautherie** [comment]
> HBelusca requested changes 

Can you update that status? Thanks.

File: boot/freeldr/freeldr/lib/peloader.c
Comment (HBelusca) [inline boot/freeldr/freeldr/lib/peloader.c:633]: might be moved just after the FrLdrHeapAlloc + NULL-check above.

Comment (SergeGautherie) [reply boot/freeldr/freeldr/lib/peloader.c:633]: If moved, that could be up to the allocation itself, for simplicity.
However, I would prefer to have it here, to emphasize that `Buffer` will change after that point.

File: boot/freeldr/freeldr/ntldr/wlregistry.c
Comment (HBelusca) [inline boot/freeldr/freeldr/ntldr/wlregistry.c:485]:
```suggestion
```

File: boot/freeldr/freeldr/ntldr/wlregistry.c
Comment (HBelusca) [inline boot/freeldr/freeldr/ntldr/wlregistry.c:548]:
```suggestion
        TRACE_CH(REACTOS, "Failed to query the 'List' value (rc %d)\n", (int)rc);
        FrLdrHeapFree(GroupNameBuffer, TAG_WLDR_NAME);
```

File: boot/freeldr/freeldr/ntldr/wlregistry.c
Comment (HBelusca) [inline boot/freeldr/freeldr/ntldr/wlregistry.c:542]:
```suggestion
        TRACE_CH(REACTOS, "Failed to allocate buffer\n");
        return;
```

File: boot/freeldr/freeldr/lib/peloader.c
Comment (HBelusca) [inline boot/freeldr/freeldr/lib/peloader.c:636]:
```suggestion
    Length /= sizeof(WCHAR);
    while (Length--)
```

File: boot/freeldr/freeldr/lib/peloader.c
Comment (HBelusca) [inline boot/freeldr/freeldr/lib/peloader.c:657]:
```suggestion
    Length /= sizeof(WCHAR);
    while (Length--)
```

File: boot/freeldr/freeldr/lib/peloader.c
Comment (HBelusca) [inline boot/freeldr/freeldr/lib/peloader.c:627]:
```suggestion

    /* Save Buffer, in case of later failure */
    BaseDllNameBuffer = Buffer;

```

File: boot/freeldr/freeldr/lib/peloader.c
Comment (HBelusca) [inline boot/freeldr/freeldr/lib/peloader.c:633]:
```suggestion
```

**Review by HBelusca — COMMENTED** [review]
Fixes.

**Comment by HBelusca** [comment]
Manually merged in master.

**Pull request closed without merging**
